src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php

```diff
@@ -2059,7 +2077,7 @@ protected function transformModelValue($key, $value)
         // If the attribute exists within the cast array, we will convert it to
         // an appropriate native PHP type dependent upon the associated value
         // given with the key in the pair. Dayle made this comment line up.
-        if ($this->hasCast($key)) {
+        if (! $value instanceof Expression && $this->hasCast($key)) {
```

Comment:
This change also makes sure that expressions work with casts (i.e. Expression objects are always used if assigned).